bastionzero/bzero
# Add support for known_hosts

## Description of the change

  - **What problem does this solve?** 
    - tl;dr: our implementation of ssh involves a lot of host key changes that we need to hide from the user's ssh process, which is ordinarily very strict about such things (even though checking host keys doesn't really do much)
  - **How does this solve that?**
    - When the user does `zli generate sshConfig`, we now add a special field that instructs ssh to use a custom known_hosts file, bz-known_hosts, instead of the default
    - When the user performs an ssh proxy via ZLI, ZLI passes the daemon a set of hostNames (one of which the user almost certainly used) and the path to bz-known_hosts\
    - The next depends on whether we are tunneling with opaque ssh or copying with transparent ssh:
      - For transparent ssh, the host key is the same as the public key of the identityfile, meaning the daemon generates it if not there or else just reads that file and retrieves the key
      - For opaque ssh, the agent will send the daemon a stream message with the remote machine's host key
    - In either case, the host key is written to bz-known_hosts before the ssh handshake is complete, so the local ssh process checks it and sees that there is a key there. Hooray!
  - **How do I test it?**
    - Once you check out this branch and pull the latest feat/transparent-ssh branch on zli, rsync your agent, and do a fresh `npm run release` / `zli generate sshConfig`, you should be able to connect via transparent or opaque ssh, perform `zli logout` / `zli login`, and never see any warnings about keys being added to known_hosts!!
  - **Any gotchas?**
    - Right now the host key writer completely overwrites the file with each new key. I can't think of a reason this would be a problem and it greatly simplifies the code, but just wanted to flag it
    - The agent only checks one location for a host key, and it only checks for an RSA formatted key while there are probably others in that spot. I think this is fine but if we wanted to, we could have it read any `*_key.pub` file it finds and send it back. **I don't think we need to do this for now but it could be that users of exotic OSs will need it**

## Testing

Describe how to test this PR....

**backend branch:** feat/transparent-ssh
**zli branch:** feat/transparent-ssh

#### Ready to run system tests?


## Relevant release note information

Release Notes:

## Related JIRA tickets

Relates to JIRA: CWC-1631

## Have you considered the security impacts?

Does this PR have any security impact?


If yes, please explain:
